File: EPI/17.3/swift/17.3.swift
public func traversalsCount(rows: Int, cols: Int) -> Int {

    // counts[row][col] is the number of ways to reach the cell at (row, col).
    var counts = [[Int]](repeating: [Int](repeating: 0, count: cols), count: rows)

    // Base cases: there is only one way to reach the cells along the top or left edge.
    for col in 0..<cols {
        counts[0][col] = 1
    }
    for row in 0..<rows {
        counts[row][0] = 1
    }

    // The cell at (row, col) can be reached either by:
    // - Moving down from the cell at (row-1, col)
    // - Moving right from the cell at (row, col-1)
    for row in 1..<rows {
        for col in 1..<cols {
            counts[row][col] = counts[row - 1][col] + counts[row][col - 1]
        }
    }

    return counts.last!.last!
}

File: EPI/17.3/typescript/17.3.ts
export default function countTraversals(rows: number, cols: number): number {
    // cache[r][c] is the traversal count for a 2D array with r rows and c columns.
    const cache = Array.from({length: rows + 1}, () => new Array(cols + 1).fill(0));

    // Base case: 2D arrays with only one row or column have only one possible traversal.
    for (let r = 1; r <= rows; r += 1) {
        cache[r][1] = 1;
    }
    for (let c = 1; c <= cols; c += 1) {
        cache[1][c] = 1;
    }

    return countTraversalsRec(rows, cols, cache);
}

function countTraversalsRec(rows: number, cols: number, cache: number[][]): number {
    const cached = cache[rows][cols];

    if (cached > 0) {
        return cached;
    }

    const result = countTraversalsRec(rows - 1, cols, cache) +
        countTraversalsRec(rows, cols - 1, cache);

    cache[rows][cols] = result;

    return result;
}
